In [16]:
import pandas as pd

In [17]:
# importing required dataset - u.data
# cols = ['userId', 'movie id', 'rating', 'timestamp']
df = pd.read_csv(r'F:\Harshi College\MS DA\Machine Learning\lab\unsupervised learning\diabetes.csv',sep=',')
print(df.head)
print(df.columns.values)

<bound method NDFrame.head of      Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50 

In [18]:
from sklearn.model_selection import train_test_split

X = df.drop(['Outcome'], axis = 1)
y = df['Outcome']

training_data, testing_data, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=5)

print(f"No. of training examples: {training_data.shape[0]}")
print(f"No. of testing examples: {testing_data.shape[0]}")

No. of training examples: 614
No. of testing examples: 154


In [19]:
training_data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
140            3      128             78              0        0  21.1   
358           12       88             74             40       54  35.3   
143           10      108             66              0        0  32.4   
674            8       91             82              0        0  35.6   
121            6      111             64             39        0  34.2   
..           ...      ...            ...            ...      ...   ...   
73             4      129             86             20      270  35.1   
400            4       95             64              0        0  32.0   
118            4       97             60             23        0  28.2   
701            6      125             78             31        0  27.6   
206            8      196             76             29      280  37.5   

     DiabetesPedigreeFunction  Age  
140                     0.268   55  
358                     0.378   48  
143                     0.272   42  
674                     0.587   68  
121                     0.260   24  
..                        ...  ...  
73                      0.231   23  
400                     0.161   31  
118                     0.443   22  
701                     0.565   49  
206                     0.605   57  

[614 rows x 8 columns]

##### KNN classifier

In [20]:
# KNN classifier

from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)

accList = []
# Import necessary modules
from sklearn.neighbors import KNeighborsClassifier
for i in range(40,50):
    knn = KNeighborsClassifier(n_neighbors=i)

    knn.fit(training_data, y_train)

    # Predict on dataset which model has not seen before
    y_pred = knn.predict(testing_data)
#     print(y_pred)

    import numpy as np
    # converting series to ndarray
    y_test = np.asarray(y_test)

    # confusion matrix
    from sklearn.metrics import confusion_matrix
    tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#     print(confusion_matrix(y_test, y_pred))
    print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

    # error evaluation methods
    # calculating accuracy

    from sklearn.metrics import classification_report
    from sklearn.metrics import accuracy_score

    #Check precision, recall, f1-score
    print(classification_report(y_test, y_pred) )

    print("accuracy_score",accuracy_score(y_test, y_pred))
    print("is with k = ",i)
    accList.append(accuracy_score(y_test, y_pred))
    print("=======================================================================================")

from statistics import mean
print("mean of the accuracy is")
mean(accList)


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:36:47
true positive : 96
 true negative : 4
 false positive : 34
 false negative : 20

              precision    recall  f1-score   support

           0       0.74      0.96      0.83       100
           1       0.83      0.37      0.51        54

    accuracy                           0.75       154
   macro avg       0.79      0.67      0.67       154
weighted avg       0.77      0.75      0.72       154

accuracy_score 0.7532467532467533
is with k =  40
true positive : 96
 true negative : 4
 false positive : 33
 false negative : 21

              precision    recall  f1-score   support

           0       0.74      0.96      0.84       100
           1       0.84      0.39      0.53        54

    accuracy                           0.76       154
   macro avg       0.79      0.67      0.69       154
weighted avg       0.78      0.76      0.73       154

accuracy_score 0.7597402597402597
is with k =  41
true positive : 98
 true negative : 2
 false positive : 34
 fa

In [21]:
from sklearn.metrics import multilabel_confusion_matrix
multilabel_confusion_matrix(y_test, y_pred)

array([[[23, 31],
        [ 6, 94]],

       [[94,  6],
        [31, 23]]], dtype=int64)

##### decision tree

In [22]:
# https://machinelearningknowledge.ai/decision-tree-classifier-in-python-sklearn-with-example/#:~:text=Decision%20tree%20is%20a%20type,internal%20nodes%2C%20and%20leaf%20nodes.

In [23]:
# Decision tree

from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)

from sklearn.tree import DecisionTreeClassifier
clf_model = DecisionTreeClassifier(criterion="gini", random_state=42, max_depth=3, min_samples_leaf=5)   
clf_model.fit(training_data,y_train)

y_pred = clf_model.predict(testing_data)
y_pred

# confusion matrix
from sklearn.metrics import confusion_matrix
tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#print(confusion_matrix(y_test, y_pred))
print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

# error evaluation methods
# calculating accuracy

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#Check precision, recall, f1-score
print(classification_report(y_test, y_pred) )

print("accuracy_score",accuracy_score(y_test, y_pred))


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:36:47
true positive : 94
 true negative : 6
 false positive : 31
 false negative : 23

              precision    recall  f1-score   support

           0       0.75      0.94      0.84       100
           1       0.79      0.43      0.55        54

    accuracy                           0.76       154
   macro avg       0.77      0.68      0.69       154
weighted avg       0.77      0.76      0.74       154

accuracy_score 0.7597402597402597
end time = 23:36:47
0:00:00


In [24]:
# Naïve Bayes classifier

from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)

from sklearn.naive_bayes import MultinomialNB

naive_bayes_classifier = MultinomialNB()
naive_bayes_classifier.fit(training_data, y_train)

y_pred = naive_bayes_classifier.predict(testing_data)
len(y_pred)

# confusion matrix
from sklearn.metrics import confusion_matrix
tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#print(confusion_matrix(y_test, y_pred))
print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

# error evaluation methods
# calculating accuracy

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#Check precision, recall, f1-score
print(classification_report(y_test, y_pred) )

print("accuracy_score",accuracy_score(y_test, y_pred))


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:36:47
true positive : 72
 true negative : 28
 false positive : 32
 false negative : 22

              precision    recall  f1-score   support

           0       0.69      0.72      0.71       100
           1       0.44      0.41      0.42        54

    accuracy                           0.61       154
   macro avg       0.57      0.56      0.56       154
weighted avg       0.60      0.61      0.61       154

accuracy_score 0.6103896103896104
end time = 23:36:47
0:00:00


In [25]:
# Logistic regression

from datetime import datetime
now = datetime.now()
start_time = now.strftime("%H:%M:%S")
print("start time =", start_time)

# Fitting Logistic Regression to the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0, solver='lbfgs', multi_class='auto')
classifier.fit(training_data, y_train)

# Predicting the Test set results
y_pred = classifier.predict(testing_data)
y_pred

# confusion matrix
from sklearn.metrics import confusion_matrix
tp_tn, fp_fn = confusion_matrix(y_test, y_pred)
#print(confusion_matrix(y_test, y_pred))
print(f"true positive : {list(tp_tn)[0]}\n true negative : {list(tp_tn)[1]}\n false positive : {list(fp_fn)[0]}\n false negative : {list(fp_fn)[1]}\n")

# error evaluation methods
# calculating accuracy

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#Check precision, recall, f1-score
print(classification_report(y_test, y_pred) )

print("accuracy_score",accuracy_score(y_test, y_pred))


now = datetime.now()
end_time = now.strftime("%H:%M:%S")
print("end time =", end_time)

# computing time complexity
time_1 = datetime.strptime(start_time,"%H:%M:%S")
time_2 = datetime.strptime(end_time,"%H:%M:%S")

time_interval = time_2 - time_1
print(time_interval)
 

start time = 23:36:47
true positive : 87
 true negative : 13
 false positive : 20
 false negative : 34

              precision    recall  f1-score   support

           0       0.81      0.87      0.84       100
           1       0.72      0.63      0.67        54

    accuracy                           0.79       154
   macro avg       0.77      0.75      0.76       154
weighted avg       0.78      0.79      0.78       154

accuracy_score 0.7857142857142857
end time = 23:36:47
0:00:00


C:\Users\Admin\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [26]:
# Linear regression


##### summary
accuracy score for pima diabetes dataset analysis:

KNN classifier - 0.7610389610389611

decision tree - 0.7597402597402597

Naive bayes classifier - 0.6103896103896104

Logistic regression - 0.7857142857142857
